server/src/main/java/org/elasticsearch/common/settings/Setting.java

```diff
@@ -1981,7 +1998,7 @@ public void diff(Settings.Builder builder, Settings source, Settings defaultSett
             if (exists(source) == false) {
                 List<String> asList = defaultSettings.getAsList(getKey(), null);
                 if (asList == null) {
-                    builder.putList(getKey(), defaultStringValue.apply(defaultSettings));
+                    builder.putList(getKey(), defaultStringValue.apply(source));
```

Comment:
I think this is a bug, but I'm not sure why we never saw it before. Maybe we never had string list settings whose default value depended on other settings before?

Reply:
Good find, I agree that the current behaviour isn't correct and actual settings need to be pushed down instead of default settings..